### PR TITLE
golangci-lint: Enable staticcheck

### LINKED ISSUE #11808: Re-enable staticcheck
We disabled staticcheck in #2629 because it was OOMing.
Since then, staticcheck migrated to the go/analysis framework
and has massively reduced its memory usage.

Per https://github.com/dominikh/go-tools/issues/376,

> where master needs 4.7 GB and 22s to check all of std, the port needs 450 MB and 17s

This makes it viable to re-enable staticcheck in CI again,
although there are a bunch of failures we'll need to fix first.

Note that if we do this via golanglint-ci,
we also have to drop megacheck from the `disable:` section because megacheck includes staticcheck,
so leaving it in there will still leave staticcheck disabled.

### PR DESCRIPTION
**Pull request by abhinav**
Remove staticcheck from the list of disabled linters.
It's enabled by default in golangci-lint.

This also fixes minor remaining staticcheck issues
that don't merit their own pull requests,
or opts out of those that cannot be fixed yet.

Notably, we're opting out of:

- Resource.Name is deprecated (#9469)
- github.com/golang/protobuf is deprecated (#11869)
- strings.Title has been deprecated (#11870)

Besides that, other issues addressed in this change are:

```
// all issues are in pkg
codegen/schema/docs_parser.go:103:4: SA4006: this value of `text` is never used (staticcheck)
codegen/schema/loader.go:253:3: SA9003: empty branch (staticcheck)
resource/deploy/step_executor.go:328:12: SA9003: empty branch (staticcheck)
resource/deploy/step_generator.go:141:10: SA9003: empty branch (staticcheck)
codegen/pcl/invoke.go:97:10: SA9003: empty branch (staticcheck)
codegen/hcl2/model/type_const.go:57:2: SA9003: empty branch (staticcheck)
codegen/hcl2/model/type_enum.go:99:9: SA4001: &*x will be simplified to x. It will not copy x. (staticcheck)
codegen/go/gen_test.go:399:19: SA4017: HasPrefix is a pure function but its return value is ignored (staticcheck)
```

Depends on #11857, #11858, #11859, #11860, #11862, #11865, #11866, #11867, #11868

Resolves #11808

---

**NOTE**: This PR's base branch is currently #11868.
The base branch will be updated when that lands.

### CODE REVIEW

File: pkg/codegen/schema/docs_parser.go [hunk 1]
```diff
@@ -100,7 +100,7 @@ func (shortcodeParser) parseShortcode(line []byte, pos int) (int, int, int, bool
 			if inName {
 				nameEnd = pos
 			}
-			text, pos = text[3:], pos+3
```
Comment (RobbieMcKinstry) [inline pkg/codegen/schema/docs_parser.go:103]: This change doesn't look like it's a refactor. 
I would expect a second assignment statement after line 103, that updates the value of `text`.

**Review by RobbieMcKinstry — APPROVED** [review]
One comment to address! Otherwise, approved! :)

File: pkg/codegen/schema/docs_parser.go (hunk 1 above)
Comment (abhinav) [reply pkg/codegen/schema/docs_parser.go:103]: Oh, it's part of the same linter fixes.
The value of `text` is never used again after this point.
If this condition matches, we set pos, break out of the loop,
and return `nameStart, nameEnd, pos, isClose, true`.
Let me add a comment here.

**Pull request merged**